src/system/process.py

```diff
@@ -0,0 +1,72 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# The OpenSearch Contributors require contributions made to
+# this file be licensed under the Apache-2.0 license or a
+# compatible open source license.
+
+import logging
+import subprocess
+import tempfile
+
+import psutil  # type: ignore
+
+
+class Process:
+    def __init__(
+        self
+    ):
+        self.process = None
+
+    def start(self, command, cwd):
+        self.stdout = tempfile.NamedTemporaryFile()
+        self.stderr = tempfile.NamedTemporaryFile()
+
+        self.process = subprocess.Popen(
+            command,
+            cwd=cwd,
+            shell=True,
+            stdout=self.stdout,
+            stderr=self.stderr,
+        )
+
+    def terminate(self):
+        parent = psutil.Process(self.process.pid)
+        logging.debug("Checking for child processes")
+        child_processes = parent.children(recursive=True)
+        for child in child_processes:
+            logging.debug(f"Found child process with pid {child.pid}")
+            if child.pid != self.process.pid:
+                logging.debug(f"Sending SIGKILL to {child.pid} ")
+                child.kill()
+        logging.info(f"Sending SIGTERM to PID {self.process.pid}")
+        self.process.terminate()
+        try:
+            logging.info("Waiting for process to terminate")
+            self.process.wait(10)
+        except subprocess.TimeoutExpired:
+            logging.info("Process did not terminate after 10 seconds. Sending SIGKILL")
+            self.process.kill()
+            try:
+                logging.info("Waiting for process to terminate")
+                self.process.wait(10)
+            except subprocess.TimeoutExpired:
+                logging.info("Process failed to terminate even after SIGKILL")
+                raise
+        finally:
+            logging.info(f"Process terminated with exit code {self.process.returncode}")
+            if self.stdout:
+                self.stdout_data = self.stdout.read()
+                self.stdout.close()
+                self.stdout = None
+            if self.stderr:
+                self.stderr_data = self.stderr.read()
+                self.stderr.close()
+                self.stderr = None
+            self.return_code = self.process.returncode
+            self.process = None
```

Comment:
Don't see the need to use terminate first, if no response then try kill.

why not just kill()

btw. we should let original call stack being bubble up, instead of `raise` a new one.  

https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it

Reply:
Let me simplify the logic to just kill()